### PR TITLE
Add SAML authentication

### PR DESCRIPTION
**Pull request by yol**
This is a basic reimplementation of #185 on current master. (related FR: #1253)

Now I know that it is not going to be merged probably, but I wanted to make this available to other interested parties so it can at least be pulled into individual installations.

### CODE REVIEW

File: config/initializers/omniauth.rb [hunk 1]
```diff
@@ -16,6 +16,9 @@
 Rails.application.config.omniauth_google = ENV['GOOGLE_OAUTH2_ID'].present? && ENV['GOOGLE_OAUTH2_SECRET'].present?
 Rails.application.config.omniauth_office365 = ENV['OFFICE365_KEY'].present? &&
                                               ENV['OFFICE365_SECRET'].present?
+Rails.application.config.omniauth_saml = ENV['SAML_ISSUER'].present? && ENV['SAML_IDP_URL'].present? &&
+                                         ENV['SAML_IDP_CERT_FINGERPRINT'].present?
+
```
Comment (einhirn) [inline config/initializers/omniauth.rb:21]: Increase the chances of getting this commited by removing this extra empty line that rubocop complains about...

Comment (yol) [reply config/initializers/omniauth.rb:21]: I hope this increases the chances by at least 100 %

**Comment by foobarable** [comment]
Thanks alot, we will most likely patch this into out setup.

**Comment by coudot** [comment]
Hello, we tested this with LemonLDAP::NG as SAML IDP and it works well. It would be great to include this in the official BBB code.

**Comment by maxbes** [comment]
I'm very interested in seeing this feature land in Greenlight.

One minor hurdle I encountered is while opening the user profile after a successful login, I get a 500 error, possibly due to language options being uninitialized

```
FATAL: [6b762aac-3125-4833-a84c-85875584f282]    
FATAL: [6b762aac-3125-4833-a84c-85875584f282] ActionView::Template::Error (undefined method `[]' for nil:NilClass): 
FATAL: [6b762aac-3125-4833-a84c-85875584f282]     36:     <%= f.text_field :provider, class: "form-control", readonly: "" %>
[6b762aac-3125-4833-a84c-85875584f282]     37:     <br>
[6b762aac-3125-4833-a84c-85875584f282]     38:     <%= f.label t("settings.account.language"), class: "form-label" %>
[6b762aac-3125-4833-a84c-85875584f282]     39: 
[6b762aac-3125-4833-a84c-85875584f282]     40:     <% current_user_role = current_user.highest_priority_role %>
[6b762aac-3125-4833-a84c-85875584f282]     41:     <br>
[6b762aac-3125-4833-a84c-85875584f282]     42:     <br> 
FATAL: [6b762aac-3125-4833-a84c-85875584f282]    
FATAL: [6b762aac-3125-4833-a84c-85875584f282] app/helpers/users_helper.rb:50:in `block in language_options'
[6b762aac-3125-4833-a84c-85875584f282] app/helpers/users_helper.rb:48:in `each'
[6b762aac-3125-4833-a84c-85875584f282] app/helpers/users_helper.rb:48:in `language_options'
[6b762aac-3125-4833-a84c-85875584f282] app/views/users/components/_account.html.erb:39:in `block in _app_views_users_components__account_html_erb___3728485878160030678_47397693265980'
[6b762aac-3125-4833-a84c-85875584f282] app/views/users/components/_account.html.erb:39:in `block in _app_views_users_components__account_html_erb___3728485878160030678_47397693265980'
[6b762aac-3125-4833-a84c-85875584f282] app/views/users/components/_account.html.erb:16:in `_app_views_users_components__setting_view_html_erb__1963868551716016331_47397693241560'
[6b762aac-3125-4833-a84c-85875584f282] app/views/users/edit.html.erb:32:in `_app_views_users_edit_html_erb__1379829543445150010_47397693054660' 
```

Thanks a lot for sharing your branch

**Comment by slater0013** [comment]
Hi,
Interested as well :) 
Hope we can see this feature soon !

**Comment by foobarable** [comment]
It won't be merged since they don't want to support it. We merged this and run it in production

**Comment by sbernhard** [comment]
This is a open source project and there were multiple attempts to add SAML2 / OpenID Connect authentications support.

Instead of adding SAML2 / OIDC, which are industry standards for doing a Single Sign On, greenlight supports Google? Twitter? Office365? authenticaton support? 

SAML2 / OIDC is supportd by many different Single Sign On components like Keycloak and its downstream product Redhat SSO. A lot of awesome tools are supporting it, too. It would be gerat to add OIDC or SAML.

**Comment by thijskh** [comment]
We are running this in production with good results. Can you please consider to merge it? It's not very invasive and as argued above because it uses open standards it can be beneficial to many different use cases.

**Comment by n0emis** [comment]
We patched this into our production setup a few months ago and it works fine for us.

**Comment by lchanouha** [comment]
+1 :)

**Comment by farhatahmad** [comment]
Now that we've merged support for OpenID Connect, there are no plans to support SAML directly. We just using OIDC providers and using SAML through the provider

**Comment by ichdasich** [comment]
@sparse91 what do you meant? The custom saml patches or the openID integration?

**Comment by sparse91** [comment]
@ichdasich This SAML patch.

**Comment by ichdasich** [comment]
You have to keep REXML at 3.2.4 (note: 3.2.5 was a security update! Assess whether it applies to your infra!) and then rebuild GL with that. 

See: 

https://github.com/omniauth/omniauth-saml/issues/199
https://github.com/onelogin/ruby-saml/issues/577

I would recommend figuring out how to migrate to the buildin openid connector, though. ;-)

**Pull request closed without merging**